### PR TITLE
REST > Added ChargingProfile get / update / delete validation

### PR DESCRIPTION
**Pull request by Ninjeneer**

### CODE REVIEW

**Comment by jerome-benoit** [comment]
@OliveGerste: since you're the smart charging expert ;), could you please review that PR and then promote it for a final review once ok? 

Thanks.

**Review by OliveGerste — CHANGES_REQUESTED** [review]
Hi Loan, good job so far.
Unfortunately we have a problem with the ChargingProfileID here. 
The validation is expecting an object ID, but the charging profiles are currently saved with a generated and hashed id from the storage method, which do not match the patterns.
Maybe we need a second opinion how to proceed here.

**Comment by Ninjeneer** [comment]
As discussed, refactored and factorized the code of create and update charging profiles

File: src/server/rest/v1/validator/ChargingStationValidator.ts [hunk 1]
```diff
@@ -140,4 +139,28 @@ export default class ChargingStationValidator extends SchemaValidator {
...
+
+  public validateChargingProfilesGetReq(data: any): HttpChargingProfilesRequest {
+    // Validate schema
+    this.validate(this.chargingProfilesGet, data);
+    return data;
+  }
+
+  public validateChargingProfileCreateReq(data: ChargingProfile): ChargingProfile {
+    // Validate schema
+    this.validate(this.chargingProfileCreate, data);
```
[the remark this answers is not in the document]
Comment (Ninjeneer) [reply src/server/rest/v1/validator/ChargingStationValidator.ts:151]: Fixed

File: src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json [hunk 2]
```diff
@@ -1,5 +1,115 @@
...
+        },
+        "transactionId": {
+          "$ref": "transaction.json#/definitions/id"
+        },
+        "stackLevel": {
+          "type": "number",
+          "sanitize": "mongo"
+        },
+        "chargingProfilePurpose": {
+          "type": "string",
```
Comment (OliveGerste) [inline src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json:30]: Enum: ChargePointMaxProfile, TxDefaultProfile, TxProfile

File: src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json [hunk 3]
```diff
@@ -1,5 +1,115 @@
...
+          "sanitize": "mongo"
+        },
+        "chargingProfileKind": {
+          "type": "string",
```
Comment (OliveGerste) [inline src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json:34]: Enum: Ansolute, Reccuring, Relative

File: src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json [hunk 4]
```diff
@@ -1,5 +1,115 @@
...
+          "sanitize": "mongo"
+        },
+        "recurrencyKind": {
+          "type": "string",
```
Comment (OliveGerste) [inline src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json:38]: Enum: Daily, Weekly

File: src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json [hunk 5]
```diff
@@ -1,5 +1,115 @@
...
+              "type": "number",
+              "sanitize": "mongo"
+            },
+            "startSchedule": {
+              "type": "string",
+              "format": "date-time",
+              "sanitize": "mongo"
+            },
+            "chargingRateUnit": {
+              "type": "string",
```
Comment (OliveGerste) [inline src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json:64]: Enum: W, A

File: src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json [hunk 6]
```diff
@@ -1,5 +1,115 @@
 {
   "title": "Create a Charging Profile",
   "type": "object",
-  "$ref": "chargingstation.json#/definitions/chargingProfile"
```
Comment (OliveGerste) [inline src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json:4]: Is it possible to reuse the profile structure from the definition somehow?
The only thing, which differs is that the id is mandatory right?

I also added some enums of the values which are possible for the different cp properties

File: src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json [hunk 7]
```diff
@@ -1,5 +1,115 @@
...
+  "properties": {
+    "chargingStationID": {
+      "$ref": "chargingstation.json#/definitions/id"
+    },
+    "connectorID": {
+      "$ref": "chargingstation.json#/definitions/connector/properties/connectorId"
+    },
+    "chargePointID": {
+      "type": "number",
```
Comment (OliveGerste) [inline src/assets/server/rest/v1/schemas/chargingstation/chargingprofile-create.json:12]: The connector ID is reused. Is there a reason, that the cchargePoint ID is not reused?

**Comment by jerome-benoit** [comment]
@OliveGerste: The status of that PR is still 'Request changes'?

**Comment by jerome-benoit** [comment]
> > @OliveGerste: The status of that PR is still 'Request changes'?
> 
> I think it is not ready yet for a review is it?

I do not know, you're its main reviewer ;)

**Pull request merged**